experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +453,184 @@ impl LocaleDisplayNamesFormatter {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
-
-        // TODO: This binary search needs to return the longest matching found prefix
-        // instead of just perfect matches
-        if let Some(displayname) = match self.options.style {
-            Some(Style::Short) => self
-                .locale_data
-                .get()
-                .short_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Long) => self
-                .locale_data
-                .get()
-                .long_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Menu) => self
-                .locale_data
-                .get()
-                .menu_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            _ => None,
-        }
-        .or_else(|| {
-            self.locale_data
-                .get()
-                .names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse())
-        }) {
-            return Cow::Borrowed(displayname);
+        let longest_matching_subtag =
+            LongestMatchingSubtag::find_longest_matching_subtag(&locale, &self);
+
+        // Step - 1: Construct a locale display name string (LDN).
+        // Find the displayname for the longest_matching_subtag which was derived above.
+        let ldn = get_locale_display_name(&locale, longest_matching_subtag, &self);
+
+        // Step - 2: Construct a vector of longest qualifying substrings (LQS).
+        // Find the displayname for the remaining locale if exists.
+        let lqs = get_longest_qualifying_substrings(&locale, longest_matching_subtag, &self);
+
+        // Step - 3: Return the displayname based on the size of LQS.
+        let mut result = Cow::Borrowed(ldn);
+        if lqs.len() > 0 {
+            let mut output = String::with_capacity(
+                result.len() + " (".len() + lqs.iter().map(|s| ", ".len() + s.len()).sum::<usize>()
+                    - ", ".len()
+                    + ")".len(),
+            );
+            output.push_str(&result);
+            output.push_str(" (");
+            output.push_str(lqs[0]);
+            for lqs in &lqs[1..] {
+                output.push_str(", ");
+                output.push_str(lqs);
+            }
+            output.push_str(")");
+            result = Cow::Owned(output);
         }
+        result
+    }
+}
 
-        // TODO: This is a dummy implementation which does not adhere to UTS35. It only uses
-        // the language and region code, and uses a hardcoded pattern to combine them.
-
-        let langdisplay = match self.options.style {
-            Some(Style::Short) => self
-                .language_data
+fn get_locale_display_name<'a>(
+    locale: &Locale,
+    longest_matching_subtag: LongestMatchingSubtag,
+    locale_dn_formatter: &'a LocaleDisplayNamesFormatter,
+) -> &'a str {
```

Comment:
Suggestion: get rid of the `'a` so that you can return references inside of `&Locale`

Reply:
When I removed it, compiler threw an error with a suggestion that lifetime is needed.